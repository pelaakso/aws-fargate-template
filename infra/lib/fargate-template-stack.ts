import { RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Repository } from 'aws-cdk-lib/aws-ecr';
import { Vpc } from 'aws-cdk-lib/aws-ec2';
import { AwsLogDriverMode, Cluster, ContainerImage, FargateTaskDefinition, LogDriver } from 'aws-cdk-lib/aws-ecs';
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';

export interface Props extends StackProps {
  /**
   * Is this stack being created as part of a unit test?
   * If so, dummy assets are used so that real code assets don't need to exist.
   *
   * @default - It is assumed that stack is not created as part of unit test and real asset path is used.
   */
  readonly isUnitTest?: boolean;

  /**
   * The VPC ID to use for the stack.
   */
  readonly vpcId: string;

  /**
   * The name used for resources where applicable.
   */
  readonly resourceName: string;

  /**
   * The amount of CPU to use for the Fargate container.
   * @default - Whatever is the default in CDK.
   */
  readonly cpu: number;

  /**
   * The amount of memory to use for the Fargate container.
   * @default - Whatever is the default in CDK.
   */
  readonly memoryLimitMiB: number;

  /**
   * The amount of time to retain CloudWatch logs.
   * @default - Whatever is the default in CDK.
   */
  readonly logRetention: RetentionDays;
}

export class FargateTemplateStack extends Stack {
  constructor(scope: Construct, id: string, props?: Props) {
    super(scope, id, props);

    const existingVpc = Vpc.fromLookup(this, 'VPC', {
      vpcId: props?.vpcId,
    });

    const ecrRepository = new Repository(this, 'FargateTemplateRepository', {
      repositoryName: props?.resourceName,
      removalPolicy: RemovalPolicy.DESTROY,
    });

    new Cluster(this, 'FargateTemplateCluster', {
      clusterName: props?.resourceName,
      enableFargateCapacityProviders: true,
      vpc: existingVpc,
    });

    const logGroup = new LogGroup(this, props?.resourceName ?? 'fargate-template', {
      logGroupName: props?.resourceName,
      retention: props?.logRetention,
    });

    const taskDef = new FargateTaskDefinition(this, 'FargateTemplateTaskDef', {
      family: props?.resourceName,
      cpu: props?.cpu,
      memoryLimitMiB: props?.memoryLimitMiB,
    });
    taskDef.addContainer('FargateTemplateContainer', {
      containerName: props?.resourceName,
      image: ContainerImage.fromEcrRepository(ecrRepository),
      logging: LogDriver.awsLogs({
        logGroup,
        streamPrefix: props?.resourceName ?? 'fargate-template',
        mode: AwsLogDriverMode.NON_BLOCKING,
      }),
    });
  }
}
