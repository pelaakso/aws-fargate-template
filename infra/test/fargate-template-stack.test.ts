import { Template } from 'aws-cdk-lib/assertions';
import * as cdk from 'aws-cdk-lib';
import { FargateTemplateStack } from '../lib/fargate-template-stack';
import { expect, test } from 'vitest';
import { RetentionDays } from 'aws-cdk-lib/aws-logs';

test('Stack snapshot', () => {
  const resourceName = 'fargate-template';
  const cpu = 512;
  const memoryLimitMiB = 1024;
  const logRetention = RetentionDays.FIVE_DAYS;

  const stack = new cdk.Stack();

  const myStack = new FargateTemplateStack(stack, 'MyTestStack', {
    env: { account: '123456789012', region: 'eu-central-1' },
    isUnitTest: true,
    vpcId: 'vpc-12345678',
    resourceName,
    cpu,
    memoryLimitMiB,
    logRetention,
  });

  const template = Template.fromStack(myStack);

  // Replace all 64 byte asset hashes, because they are likely to change from build to build.
  // For example 02eaccf2c7a5bca24a1360de04a6ec227dfbebb07d930a867f0fe8ee5fc32f4d.zip
  expect.addSnapshotSerializer({
    test: (val) => (typeof val === 'string' && val.match(/[0-9a-f]{64}.zip/) ? true : false),
    print: (val) => {
      if (typeof val === 'string') {
        return val.replace(/[0-9a-f]{64}.zip/, '64-byte-asset-hash-removed.zip');
      }
      return `${val}`;
    },
  });

  expect(template).toMatchSnapshot();
});
