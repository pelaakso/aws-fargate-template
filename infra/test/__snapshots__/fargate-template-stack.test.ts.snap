// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "FargateTemplateCluster026F9EF7": {
      "Properties": {
        "CapacityProviders": [
          "FARGATE",
          "FARGATE_SPOT",
        ],
        "Cluster": {
          "Ref": "FargateTemplateClusterF4373A98",
        },
        "DefaultCapacityProviderStrategy": [],
      },
      "Type": "AWS::ECS::ClusterCapacityProviderAssociations",
    },
    "FargateTemplateClusterF4373A98": {
      "Properties": {
        "ClusterName": "fargate-template",
      },
      "Type": "AWS::ECS::Cluster",
    },
    "FargateTemplateRepositoryC777A5F7": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "RepositoryName": "fargate-template",
      },
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Delete",
    },
    "FargateTemplateTaskDef21E463A5": {
      "Properties": {
        "ContainerDefinitions": [
          {
            "Essential": true,
            "Image": {
              "Fn::Join": [
                "",
                [
                  {
                    "Fn::Select": [
                      4,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "FargateTemplateRepositoryC777A5F7",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".dkr.ecr.",
                  {
                    "Fn::Select": [
                      3,
                      {
                        "Fn::Split": [
                          ":",
                          {
                            "Fn::GetAtt": [
                              "FargateTemplateRepositoryC777A5F7",
                              "Arn",
                            ],
                          },
                        ],
                      },
                    ],
                  },
                  ".",
                  {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/",
                  {
                    "Ref": "FargateTemplateRepositoryC777A5F7",
                  },
                  ":latest",
                ],
              ],
            },
            "LogConfiguration": {
              "LogDriver": "awslogs",
              "Options": {
                "awslogs-group": {
                  "Ref": "fargatetemplateD2D01DE7",
                },
                "awslogs-region": "eu-central-1",
                "awslogs-stream-prefix": "fargate-template",
                "mode": "non-blocking",
              },
            },
            "Name": "fargate-template",
          },
        ],
        "Cpu": "512",
        "ExecutionRoleArn": {
          "Fn::GetAtt": [
            "FargateTemplateTaskDefExecutionRoleE0DB96BD",
            "Arn",
          ],
        },
        "Family": "fargate-template",
        "Memory": "1024",
        "NetworkMode": "awsvpc",
        "RequiresCompatibilities": [
          "FARGATE",
        ],
        "TaskRoleArn": {
          "Fn::GetAtt": [
            "FargateTemplateTaskDefTaskRoleBFD9A5A3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ECS::TaskDefinition",
    },
    "FargateTemplateTaskDefExecutionRoleDefaultPolicy7A996704": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "FargateTemplateRepositoryC777A5F7",
                  "Arn",
                ],
              },
            },
            {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "fargatetemplateD2D01DE7",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FargateTemplateTaskDefExecutionRoleDefaultPolicy7A996704",
        "Roles": [
          {
            "Ref": "FargateTemplateTaskDefExecutionRoleE0DB96BD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FargateTemplateTaskDefExecutionRoleE0DB96BD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "FargateTemplateTaskDefTaskRoleBFD9A5A3": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "ecs-tasks.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "fargatetemplateD2D01DE7": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "LogGroupName": "fargate-template",
        "RetentionInDays": 5,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
